packages/langium/src/grammar/grammar-util.ts

```diff
@@ -55,24 +57,41 @@ function isDataTypeRuleInternal(rule: ast.ParserRule, visited: Set<ast.ParserRul
     return true;
 }
 
-export function findNameAssignment(rule: ast.ParserRule): ast.Assignment | undefined {
-    for (const node of streamAllContents(rule)) {
+export function findNameAssignment(type: ast.AbstractType): ast.Assignment | undefined {
+    return findNameAssignmentInternal(type, new Map());
+}
+
+function findNameAssignmentInternal(type: ast.AbstractType, cashed: Map<ast.AbstractType, ast.Assignment | undefined>): ast.Assignment | undefined {
+    function go(node: AstNode, refType: ast.AbstractType): ast.Assignment | undefined {
+        let childAssignment: ast.Assignment | undefined = undefined;
+        const parentAssignment = getContainerOfType(node, ast.isAssignment);
+        // No parent assignment implies unassigned rule call
+        if (!parentAssignment) {
+            childAssignment = findNameAssignmentInternal(refType, cashed);
+        }
+        cashed.set(type, childAssignment);
+        return childAssignment;
+    }
+
+    if (cashed.has(type)) return cashed.get(type);
+    cashed.set(type, undefined);
+    for (const node of streamAllContents(type)) {
         if (ast.isAssignment(node) && node.feature.toLowerCase() === 'name') {
+            cashed.set(type, node);
             return node;
         } else if (ast.isRuleCall(node) && ast.isParserRule(node.rule.ref)) {
-            const parentAssignment = getContainerOfType(node, ast.isAssignment);
-            // No parent assignment implies unassigned rule call
-            if (!parentAssignment) {
-                const childNameAssignment = findNameAssignment(node.rule.ref);
-                if (childNameAssignment) {
-                    return childNameAssignment;
-                }
-            }
+            return go(node, node.rule.ref);
+        } else if (ast.isAtomType(node) && node.refType.ref) {
+            return go(node, node.refType.ref);
         }
     }
     return undefined;
 }
 
+export function distictAndSorted<T>(list: T[], compareFn?: (a: T, b: T) => number): T[] {
+    return stream(list).distinct().toArray().sort(compareFn);
```

Comment:
The intermediate stream does not help here. Proposal:
```suggestion
    return Array.from(new Set(list)).sort(compareFn);
```